core/src/common.cpp

```diff
@@ -207,6 +218,143 @@ namespace irods::http
+			return _url.host();
+		}
+		return fmt::format("{}:{}", _url.host(), _port);
+	}
+
+	auto create_oidc_request(boost::urls::url_view _url) -> beast::http::request<beast::http::string_body> {
+		constexpr auto http_version_number{11};
+		beast::http::request<beast::http::string_body> req{beast::http::verb::post, _url.path(), http_version_number};
+
+		const auto port{get_port_from_url(_url)};
+
+		req.set(beast::http::field::host, create_host_field(_url, *port));
+		req.set(beast::http::field::user_agent, irods::http::version::server_name);
+		req.set(beast::http::field::content_type,
+		        "application/x-www-form-urlencoded");
+		req.set(beast::http::field::accept, "application/json");
+
+		if (const auto& secret_key{irods::http::globals::oidc_configuration().find("client_secret")}; secret_key != std::end(irods::http::globals::oidc_configuration())) {
+			const auto format_bearer_token{[](std::string_view _client_id, std::string_view _client_secret){
+				auto encode_me{fmt::format("{}:{}", encode(_client_id), encode(_client_secret))};
+				return safe_base64_encode(encode_me);
+			}};
+
+			const auto& client_id{irods::http::globals::oidc_configuration().get_ref<const std::string&>()};
+			const auto& client_secret{secret_key->get_ref<const std::string&>()};
+			const auto auth_string{fmt::format("Basic {}", format_bearer_token(client_id, client_secret))};
+
+			req.set(beast::http::field::authorization, auth_string);
+		}
+
+		return req;
+	}
+
+    auto hit_introspection_endpoint(std::string _encoded_body) -> nlohmann::json
+	{
+		const auto introspection_endpoint{
+			irods::http::globals::oidc_endpoint_configuration().at("introspection_endpoint").get_ref<const std::string&>()};
+
+		const auto parsed_uri{boost::urls::parse_uri(introspection_endpoint)};
+
+		if (parsed_uri.has_error()) {
+			log::error(
+				"{}: Error trying to parse introspection_endpoint [{}]. Please check configuration.", __func__, introspection_endpoint);
+			return {{"error", "bad endpoint"}};
+		}
+
+		const auto url{*parsed_uri};
+		const auto port{get_port_from_url(url)};
+
+		// Addr
+		net::io_context io_ctx;
+		auto tcp_stream{irods::http::transport_factory(url.scheme_id(), io_ctx)};
+		tcp_stream->connect(url.host(), *port);
+	
+		// Build Request
+		auto req{create_oidc_request(url)};
+		req.body() = std::move(_encoded_body);
+		req.prepare_payload();
+
+		// Send request & receive response
+		auto res{tcp_stream->communicate(req)};
+		
+		log::debug("{}: Received the following response: [{}]", __func__, res.body());
+
+		// JSONize response
+		return nlohmann::json::parse(res.body());
+	}
+
+	auto map_json_to_user(const nlohmann::json& _json) -> std::optional<std::string> {
+		const auto oidc_config{irods::http::globals::oidc_configuration()};
+		const static bool use_user_claim{oidc_config.contains("irods_user_claim")};
+		const static bool use_attibute_mapping{oidc_config.contains("user_attribute_mapping")};
```

Comment:
Consider making these iterators since the config is readonly.

Then the rest of the code can do a dereference without needing to do another lookup.

Reply:
Done.